### DISCUSSION TITLE
WORM / Object locking management for ransomware protection

### DISCUSSION THREAD

**Discussion by bodik** [minio/minio#19059] [Q&A]
Hello,

recently we've been considering new technologies for performing server backups. We've evaluated few existing tools like bacula, rclone and restic. Where every one of the tools has it's pros-and-cons we'r somehow inclined to use restic with s3 backends which looks scalable and somewhat modern solution. Here comes Minio as S3 provider in our testbed, hence I'd like to ask for a question(s) here regarding it's WORM usecase. 

Performing backup from backed-up resources (storage) to S3 has one significant disadvantage, the credentials has to be stored on backed-up resource. Generaly, a WORM / Object locking is advised to protect the data against ransomware attacker. Overall concept looks somewhat fine, any object stored will be locked in governance mode for let's say 100years. This should create a protection barrier which prevents an attacker from completely destroying the backups ...

... but WORM will create kind of mess in data storage consisting of many objects and their respective versions. If any attack would take place, 

* which tool or mechanism can administrator use to recreate the bucket state for some specific point in time ?
  one can perhaps do such thing manually given that metadata for any object and it's version is available,
  but it's not seem to be any practical or simple

  is there anything in minio ecosystem or tooling which would help with that task ?

* making bucket WORMed, the ammount of data will only increase in time. From time to time, it might be acceptable to
  check current state of the bucket, confirm it contains a consistent and correct backups for the given resource
  and delete all previous versions of the objects, hence reclaim some free space. this again does not seem
  like simple task.

  is there anything in minio which would facilitate such garbage collection ?

* I'm really sure that I'm not the very first person thinking like this, so any reference for such topic would be appreciated

Thanks for any response
bodik

**Comment by xmontero** [comment]
I wonder if this question should be asked in a non-minio-specific forum to get some traction.

The concepts behind can be applicable to any "controller" connecting to "any S3-compatible backend".

I am just interested in creating "some WORM" easily for my company, not decided what back will I use, nor the front, and still I feel your question is relevant to me.

**Reply by bodik** [reply]
@xmontero in the meantime I've come up with "some WORM" in https://github.com/bodik/rwm, althou it's highly opinionated and strongly tied to ceph s3 implementation (mainly because mino has very different ACLs and policies management and Ceph and AWS)
